Graylog Enterprise Report generation failing

First of all, I appreciate all the work the Graylog Community and maintainers put into it to make it a GREAT product.

In my current setup, Graylog Cluster of 3 nodes is setup as follows:

Client -> Load balancer (443) -> Nginx (https 9000) -> Graylog (https 9001)

Client browser accesses Graylog via IP on Load Balancer as url `https://graylog.example.com/` URL. This is also the `http_external_uri`. `graylog.example.com` is not a DNS entry that server can resolve but it works on the client's machines.

Then there's `http_publish_uri` set to `https://<NodeIP>:9001/`.

All of the functionality works as expected. However after installation of the Graylog Small Business Enterprise License for the reporting feature, whenever I try to generate report through the UI, I get following error:

```
2021-03-15T17:05:29.672-04:00 ERROR [ReportRenderJob] Error during report generation: com.github.rholder.retry.RetryException: Retrying failed to complete successfully after 3 attempts. - browser logs: 
2021-03-15T17:05:29.672-04:00 ERROR [ReportRenderJob] Failed to generate report: 
com.github.rholder.retry.RetryException: Retrying failed to complete successfully after 3 attempts.
        at com.github.rholder.retry.Retryer.call(Retryer.java:174) ~[graylog.jar:?]
        at org.graylog.plugins.report.scheduler.ReportRenderJob.doRun(ReportRenderJob.java:74) [graylog-plugin-enterprise-3.3.9.jar:?]
        at org.graylog.plugins.report.scheduler.ReportRenderJob.run(ReportRenderJob.java:60) [graylog-plugin-enterprise-3.3.9.jar:?]
        at org.graylog.plugins.report.scheduler.ReportRenderSystemJob.execute(ReportRenderSystemJob.java:25) [graylog-plugin-enterprise-3.3.9.jar:?]
        at org.graylog2.system.jobs.SystemJobManager$1.run(SystemJobManager.java:89) [graylog.jar:?]
        at com.codahale.metrics.InstrumentedScheduledExecutorService$InstrumentedRunnable.run(InstrumentedScheduledExecutorService.java:241) [graylog.jar:?]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) [?:1.8.0_282]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) [?:1.8.0_282]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180) [?:1.8.0_282]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293) [?:1.8.0_282]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_282]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_282]
        at java.lang.Thread.run(Thread.java:748) [?:1.8.0_282]
Caused by: org.graylog.plugins.report.rest.chromedriver.ChromeDriverException: Render engine is unable to fully generate requested report, aborting. Browser logs: 
        [BROWSER]: [2021-03-15T17:02:29-0400] [SEVERE] https://graylog.example.com/assets/vendor.8d6aa5835c8302c41e2f.js - Failed to load resource: net::ERR_NAME_NOT_RESOLVED
        [BROWSER]: [2021-03-15T17:02:29-0400] [SEVERE] https://graylog.example.com/assets/polyfill.7b2d519528c4e87cbc34.js - Failed to load resource: net::ERR_NAME_NOT_RESOLVED
        [BROWSER]: [2021-03-15T17:02:29-0400] [SEVERE] https://graylog.example.com/assets/plugin/org.graylog.plugins.collector.CollectorPlugin/plugin.org.graylog.plugins.collector.CollectorPlugin.0f2e4cf8cefdf620a71d.js - Failed to load resource: net::ERR_NAME_NOT_RESOLVED
        [BROWSER]: [2021-03-15T17:02:29-0400] [SEVERE] https://graylog.example.com/assets/plugin/org.graylog.plugins.threatintel.ThreatIntelPlugin/plugin.org.graylog.plugins.threatintel.ThreatIntelPlugin.2148ca040de9a7a73c28.js - Failed to load resource: net::ERR_NAME_NOT_RESOLVED
        [BROWSER]: [2021-03-15T17:02:29-0400] [SEVERE] https://graylog.example.com/assets/builtins.7b2d519528c4e87cbc34.js - Failed to load resource: net::ERR_NAME_NOT_RESOLVED
        [BROWSER]: [2021-03-15T17:02:29-0400] [SEVERE] https://graylog.example.com/assets/plugin/org.graylog.plugins.enterprise.EnterprisePlugin/plugin.org.graylog.plugins.enterprise.EnterprisePlugin.2d424d9fac3bf38109fc.js - Failed to load resource: net::ERR_NAME_NOT_RESOLVED
        [BROWSER]: [2021-03-15T17:02:29-0400] [SEVERE] https://graylog.example.com/assets/app.7b2d519528c4e87cbc34.js - Failed to load resource: net::ERR_NAME_NOT_RESOLVED
        [BROWSER]: [2021-03-15T17:02:29-0400] [SEVERE] https://graylog.example.com/config.js - Failed to load resource: net::ERR_NAME_NOT_RESOLVED
        [BROWSER]: [2021-03-15T17:02:29-0400] [SEVERE] https://graylog.example.com/assets/plugin/org.graylog.aws.AWSPlugin/plugin.org.graylog.aws.AWSPlugin.0898b5f648fe40b44913.js - Failed to load resource: net::ERR_NAME_NOT_RESOLVED
        at org.graylog.plugins.report.api.ReportService.fetchPdf(ReportService.java:260) ~[?:?]
        at org.graylog.plugins.report.scheduler.ReportRenderJob.lambda$doRun$0(ReportRenderJob.java:64) ~[?:?]
        at com.github.rholder.retry.AttemptTimeLimiters$NoAttemptTimeLimit.call(AttemptTimeLimiters.java:78) ~[graylog.jar:?]
        at com.github.rholder.retry.Retryer.call(Retryer.java:160) ~[graylog.jar:?]
        ... 12 more
```

According to the doc as I understood I need to adjust the `report_render_uri` which I have set to `https://<NodeIP>:9001/`. 

Now I can solve this problem by adding entry for `graylog.example.com` in `/etc/hosts` file but and I also want to know if `report_render_uri` config if its working as expected or not? Did I misconfigure anything?

Here's Nginx configuration for Graylog. Not sure if that makes a difference.

```
server {
    listen 9000 ssl;

    server_name graylog.example.com;



    root /usr/share/nginx/html;

    index index.html index.htm;

    access_log /var/log/nginx/graylog_example_com-access.log;
    error_log /var/log/nginx/graylog_example_com-error.log error;


    ssl_certificate_key /etc/nginx/tls/key.pem;
    ssl_certificate /etc/nginx/tls/cert_chain.pem;

    location / {
      proxy_read_timeout  900;
      proxy_set_header    Host $http_host;
      proxy_set_header    X-Forwarded-Host $host;
      proxy_set_header    X-Forwarded-Server $host;
      proxy_set_header    X-Forwarded-For $proxy_add_x_forwarded_for;
      proxy_pass          https://127.0.0.1:9001;
    }

}
```

## Expected Behavior
Report plugin to use `report_render_uri` for report generation.

## Current Behavior
Report plugin seems to use `http_external_uri` for report generation.

## Possible Solution


## Steps to Reproduce (for bugs)
1. Install Graylog Open source.
2. Obtain Small Business License
3. Install plugin package
4. Setup reporting following the doc https://docs.graylog.org/en/3.3/pages/reporting/setup.html
5. Setup a Test report.
6. Click on `Send Report Now` to trigger generation of the particular report to be emailed after completion.

## Context
Trying to configure scheduled reports to be delivered via email.

## Your Environment
* Graylog Version: 3.3.9
* Java Version: java-1.8.0-openjdk-headless-1.8.0.282.b08-1.el7_9.x86_64
* Elasticsearch Version: 5.6.16
* MongoDB Version: 3.6.21
* Operating System: RHEL 7.9
* Browser version: Firefox 86 on Ubuntu Linux 20.04.2

--- a/graylog2-server/src/main/java/org/graylog2/rest/RestTools.java
+++ b/graylog2-server/src/main/java/org/graylog2/rest/RestTools.java
@@ -120,6 +120,12 @@ public class RestTools {
         return uri;
     }
 
+    public static URI buildRelativeExternalUri(@NotNull MultivaluedMap<String, String> httpHeaders, @NotNull URI defaultUri) {
+        final URI externalUri = RestTools.buildExternalUri(httpHeaders, defaultUri);
+        final String path = externalUri.getPath() + (externalUri.getPath().endsWith("/") ? "" : "/");
+        return URI.create(path);
+    }
+
     public static String getPathFromResource(Resource resource) {
         String path = resource.getPath();
         Resource parent = resource.getParent();

--- a/graylog2-server/src/main/java/org/graylog2/rest/RestTools.java
+++ b/graylog2-server/src/main/java/org/graylog2/rest/RestTools.java
@@ -122,8 +122,7 @@ public class RestTools {
 
     public static URI buildRelativeExternalUri(@NotNull MultivaluedMap<String, String> httpHeaders, @NotNull URI defaultUri) {
         final URI externalUri = RestTools.buildExternalUri(httpHeaders, defaultUri);
-        final String path = externalUri.getPath() + (externalUri.getPath().endsWith("/") ? "" : "/");
-        return URI.create(path);
+        return URI.create(externalUri.getPath());
     }
 
     public static String getPathFromResource(Resource resource) {

--- a/graylog2-server/src/main/java/org/graylog2/shared/rest/resources/documentation/DocumentationBrowserResource.java
+++ b/graylog2-server/src/main/java/org/graylog2/shared/rest/resources/documentation/DocumentationBrowserResource.java
@@ -37,6 +37,7 @@ import javax.ws.rs.core.HttpHeaders;
 import javax.ws.rs.core.MediaType;
 import javax.ws.rs.core.Response;
 import java.io.IOException;
+import java.net.URI;
 import java.net.URL;
 import java.nio.charset.StandardCharsets;
 import java.util.Map;
@@ -89,8 +90,10 @@ public class DocumentationBrowserResource extends RestResource {
     public String allIndex(@Context HttpHeaders httpHeaders) throws IOException {
         final URL templateUrl = this.getClass().getResource("/swagger/index.html.template");
         final String template = Resources.toString(templateUrl, StandardCharsets.UTF_8);
+        final URI relativePath = RestTools.buildRelativeExternalUri(httpHeaders.getRequestHeaders(), httpConfiguration.getHttpExternalUri());
+
         final Map<String, Object> model = ImmutableMap.of(
-                "baseUri", RestTools.buildExternalUri(httpHeaders.getRequestHeaders(), httpConfiguration.getHttpExternalUri()).resolve(HttpConfiguration.PATH_API).toString(),
+                "baseUri", relativePath.resolve(HttpConfiguration.PATH_API).toString(),
                 "globalModePath", "global/index.html",
                 "globalUriMarker", "/global",
                 "showWarning", "true");

--- a/graylog2-server/src/main/java/org/graylog2/shared/rest/resources/documentation/DocumentationResource.java
+++ b/graylog2-server/src/main/java/org/graylog2/shared/rest/resources/documentation/DocumentationResource.java
@@ -109,7 +109,7 @@ public class DocumentationResource extends RestResource {
         URI baseUri;
         if (route.startsWith("global")) {
             route = route.replace("global", "");
-            baseUri = RestTools.buildExternalUri(httpHeaders.getRequestHeaders(), httpConfiguration.getHttpExternalUri()).resolve(HttpConfiguration.PATH_API);
+            baseUri = RestTools.buildRelativeExternalUri(httpHeaders.getRequestHeaders(), httpConfiguration.getHttpExternalUri()).resolve(HttpConfiguration.PATH_API);
         } else {
             baseUri = httpConfiguration.getHttpPublishUri().resolve(HttpConfiguration.PATH_API);
         }

--- a/graylog2-server/src/main/java/org/graylog2/web/DevelopmentIndexHtmlGenerator.java
+++ b/graylog2-server/src/main/java/org/graylog2/web/DevelopmentIndexHtmlGenerator.java
@@ -28,6 +28,7 @@ import javax.inject.Inject;
 import javax.inject.Singleton;
 import javax.ws.rs.core.MultivaluedMap;
 import java.io.IOException;
+import java.net.URI;
 import java.nio.charset.StandardCharsets;
 import java.util.Map;
 
@@ -52,9 +53,10 @@ public class DevelopmentIndexHtmlGenerator implements IndexHtmlGenerator {
 
     @Override
     public String get(MultivaluedMap<String, String> headers) {
+        final URI relativePath = RestTools.buildRelativeExternalUri(headers, httpConfiguration.getHttpExternalUri());
         final Map<String, Object> model = ImmutableMap.<String, Object>builder()
                 .put("title", "Graylog DEVELOPMENT Web Interface")
-                .put("appPrefix", RestTools.buildExternalUri(headers, httpConfiguration.getHttpExternalUri()))
+                .put("appPrefix", relativePath)
                 .build();
         return templateEngine.transform(getTemplate(), model);
     }

--- a/graylog2-server/src/main/java/org/graylog2/web/ProductionIndexHtmlGenerator.java
+++ b/graylog2-server/src/main/java/org/graylog2/web/ProductionIndexHtmlGenerator.java
@@ -26,6 +26,7 @@ import javax.inject.Inject;
 import javax.inject.Singleton;
 import javax.ws.rs.core.MultivaluedMap;
 import java.io.IOException;
+import java.net.URI;
 import java.nio.charset.StandardCharsets;
 import java.util.List;
 import java.util.Map;
@@ -72,11 +73,14 @@ public class ProductionIndexHtmlGenerator implements IndexHtmlGenerator {
 
     @Override
     public String get(MultivaluedMap<String, String> headers) {
+        final URI externalUri = RestTools.buildExternalUri(headers, httpConfiguration.getHttpExternalUri());
+        final String path = externalUri.getPath() + (externalUri.getPath().endsWith("/") ? "" : "/");
+        final URI relativePath = URI.create(path);
         final Map<String, Object> model = ImmutableMap.<String, Object>builder()
                 .put("title", "Graylog Web Interface")
                 .put("cssFiles", cssFiles)
                 .put("jsFiles", sortedJsFiles)
-                .put("appPrefix", RestTools.buildExternalUri(headers, httpConfiguration.getHttpExternalUri()))
+                .put("appPrefix", relativePath)
                 .build();
         return templateEngine.transform(template, model);
     }

--- a/graylog2-server/src/main/java/org/graylog2/web/ProductionIndexHtmlGenerator.java
+++ b/graylog2-server/src/main/java/org/graylog2/web/ProductionIndexHtmlGenerator.java
@@ -73,9 +73,7 @@ public class ProductionIndexHtmlGenerator implements IndexHtmlGenerator {
 
     @Override
     public String get(MultivaluedMap<String, String> headers) {
-        final URI externalUri = RestTools.buildExternalUri(headers, httpConfiguration.getHttpExternalUri());
-        final String path = externalUri.getPath() + (externalUri.getPath().endsWith("/") ? "" : "/");
-        final URI relativePath = URI.create(path);
+        final URI relativePath = RestTools.buildRelativeExternalUri(headers, httpConfiguration.getHttpExternalUri());
         final Map<String, Object> model = ImmutableMap.<String, Object>builder()
                 .put("title", "Graylog Web Interface")
                 .put("cssFiles", cssFiles)

--- a/graylog2-server/src/main/java/org/graylog2/web/resources/AppConfigResource.java
+++ b/graylog2-server/src/main/java/org/graylog2/web/resources/AppConfigResource.java
@@ -74,7 +74,7 @@ public class AppConfigResource {
             throw new RuntimeException("Unable to read AppConfig template while generating web interface configuration: ", e);
         }
 
-        final URI baseUri = RestTools.buildExternalUri(headers.getRequestHeaders(), httpConfiguration.getHttpExternalUri());
+        final URI baseUri = RestTools.buildRelativeExternalUri(headers.getRequestHeaders(), httpConfiguration.getHttpExternalUri());
         final Map<String, Object> model = ImmutableMap.of(
             "rootTimeZone", configuration.getRootTimeZone(),
             "serverUri", baseUri.resolve(HttpConfiguration.PATH_API),

--- a/graylog2-server/src/main/resources/swagger/lib/shred.bundle.js
+++ b/graylog2-server/src/main/resources/swagger/lib/shred.bundle.js
@@ -858,11 +858,19 @@ Object.defineProperties(Request.prototype, {
 
   url: {
     get: function() {
-      if (!this.scheme) { return null; }
-      return sprintf("%s://%s:%s%s",
-          this.scheme, this.host, this.port,
-          (this.proxy ? "/" : this.path) +
-          (this.query ? ("?" + this.query) : ""));
+      let result = '';
+      if (this.scheme) {
+        result += this.scheme + "://";
+      }
+
+      if (this.host) {
+        result += this.host;
+      }
+
+      if (this.port) {
+        result += ':' + this.port;
+      }
+      return result + (this.proxy ? "/" : this.path) + (this.query ? ("?" + this.query) : "");
     },
     set: function(_url) {
       _url = parseUri(_url);
@@ -895,13 +903,13 @@ Object.defineProperties(Request.prototype, {
 
   port: {
     get: function() {
-      if (!this._port) {
+      /*if (!this._port) {
         switch(this.scheme) {
           case "https": return this._port = 443;
           case "http":
           default: return this._port = 80;
         }
-      }
+      }*/
       return this._port;
     },
     set: function(value) { this._port = value; return this; },

--- a/graylog2-server/src/main/resources/swagger/lib/shred.bundle.js
+++ b/graylog2-server/src/main/resources/swagger/lib/shred.bundle.js
@@ -903,13 +903,6 @@ Object.defineProperties(Request.prototype, {
 
   port: {
     get: function() {
-      /*if (!this._port) {
-        switch(this.scheme) {
-          case "https": return this._port = 443;
-          case "http":
-          default: return this._port = 80;
-        }
-      }*/
       return this._port;
     },
     set: function(value) { this._port = value; return this; },

